### PR TITLE
Add user account

### PR DESCRIPTION
**Pull request by hargup**

### CODE REVIEW

File: src/clj/doit/auth/auth.clj [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+  (:require
+   [org.httpkit.client :as http]
+   [doit.user.db :as user-db]
+   [doit.config :as config]
+   [clojure.data.json :as json]))
+
+(defn token-info-url [token]
+  (format "https://www.googleapis.com/oauth2/v3/tokeninfo?id_token=%s" token))
+
+(defn current-unix-time []
```
Comment (kirang89) [inline src/clj/doit/auth/auth.clj]: This can be moved into utils namespace.

File: src/clj/doit/auth/auth.clj [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+  (quot (System/currentTimeMillis) 1000))
+
+(defn not-expired? [exp]
```
Comment (kirang89) [inline src/clj/doit/auth/auth.clj]: This can be moved into utils namespace.

File: src/clj/doit/auth/route.clj [hunk 3]
```diff
@@ -0,0 +1,6 @@
+(ns doit.auth.route
+  (:require [doit.auth.handler :as handler]
+            [doit.middleware :as mw]))
+
+(def route {"client-id/" {:get handler/client-id}
```
Comment (kirang89) [inline src/clj/doit/auth/route.clj]: should be `routes`

File: src/clj/doit/middleware.clj [hunk 4]
```diff
@@ -0,0 +1,37 @@
+(ns doit.middleware
+  (:require
+   [doit.auth.auth :as auth]
+   [clojure.string :as str]
+   [clojure.data.json :as json]))
+
+(defn forbidden-response []
```
Comment (kirang89) [inline src/clj/doit/middleware.clj]: This should be in an `error` namespace.

File: src/clj/doit/route.clj [hunk 5]
```diff
@@ -1,4 +1,7 @@
 (ns doit.route
-  (:require [doit.todo.route :as todo-route]))
+  (:require [doit.todo.route :as todo-route]
+            [doit.auth.route :as auth-route]
+            [doit.middleware :as mw]))
 
-(def route ["/api/" {"todo/" todo-route/route}])
+(def route ["/api/" {"todo/" todo-route/route
```
Comment (kirang89) [inline src/clj/doit/route.clj]: should be `routes`.

File: src/cljs/doit/auth.cljs [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+            [doit.events :as events]
+            [re-frame.core :as rf]
+            [goog.object]))
+
+
+(defn get-auth-instance []
+  (.getAuthInstance (goog.object/get js/gapi "auth2")))
+
+(defn save-auth-token
+  [cofx [_ token]]
```
Comment (kirang89) [inline src/cljs/doit/auth.cljs]: Destructure `db` directly from `cofx`.

File: src/cljs/doit/auth.cljs [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+  {:db (assoc-in (:db cofx) [:user :token] token)
+   :dispatch [::events/get-todos]})
+
+(rf/reg-event-fx
+ ::save-auth-token
+ save-auth-token)
+
+(defn remove-user-details
+  [cofx _]
```
Comment (kirang89) [inline src/cljs/doit/auth.cljs]: Destructure `db` directly from `cofx`.

File: src/cljs/doit/auth.cljs [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+(defn sign-in-failure []
+  (prn "sign in failed"))
+
+(defn sign-in []
+  (-> (.signIn (get-auth-instance))
+      (.then sign-in-success sign-in-failure)))
+
+(defn sign-out []
+  (-> (.signOut (get-auth-instance))
+      (.then #(rf/dispatch [::events/initialize-db]))))
```
Comment (kirang89) [inline src/cljs/doit/auth.cljs]: You should instead dispatch `::events/signed-out` so that you can decouple specific events(like `initialize-db`) from `auth`.

File: src/cljs/doit/config.cljs [hunk 9]
```diff
@@ -2,3 +2,5 @@
 
 (def debug?
   ^boolean goog.DEBUG)
+
+(def sign-in-btn-id "sign-in-btn")
```
Comment (kirang89) [inline src/cljs/doit/config.cljs]: Why is this required?

File: src/cljs/doit/events.cljs [hunk 10]
```diff
@@ -21,6 +21,33 @@
      :after (fn [context]
               (check-db-spec context (get-in context [:effects :db]))))))
 
+(defn token-headers-map [cofx]
```
Comment (kirang89) [inline src/cljs/doit/events.cljs]: Shouldn't this be in the `auth` namespace?

File: test/clj/doit/todo/api_test.clj [hunk 11]
```diff
@@ -1,54 +1,80 @@
 (ns doit.todo.api-test
```
Comment (kirang89) [inline test/clj/doit/todo/api_test.clj]: You should also add tests to check program behavior on sign-in success and failure.

File: src/cljs/doit/config.cljs (hunk 9 above)
Comment (hargup) [reply src/cljs/doit/config.cljs]: It is not, will remove this.

File: src/cljs/doit/events.cljs (hunk 10 above)
Comment (hargup) [reply src/cljs/doit/events.cljs]: Moving this to `auth` introduces circular dependency as `auth` dispatches events from `events`

Comment (kirang89) [reply src/cljs/doit/events.cljs]: Alright, this is ok for now.

File: test/cljs/doit/core_test.cljs [hunk 12]
```diff
@@ -30,7 +31,12 @@
          new-todo-done   {:content "new todo" :done true :id 3}
          all-todos       (rf/subscribe [::subs/todos])
          completed-todos (rf/subscribe [::subs/completed-todos])
-         remaining-todos (rf/subscribe [::subs/remaining-todos])]
+         remaining-todos (rf/subscribe [::subs/remaining-todos])
+         auth-token      (rf/subscribe [::subs/auth-token])]
+
+     (testing "signin success can set auth-token"
```
Comment (kirang89) [inline test/cljs/doit/core_test.cljs]: Here, you are testing the `save-auth-token` handler and not the sign-in flow itself. Rename this to "Can set auth token" or something appropriate.

File: src/clj/doit/routes.clj [hunk 13]
```diff
@@ -0,0 +1,7 @@
+(ns doit.routes
```
[the remark this answers is not in the document]
Comment (hargup) [reply src/clj/doit/routes.clj]: Got it.

File: src/cljs/doit/auth.cljs [hunk 14]
```diff
@@ -37,16 +28,25 @@
...
 (defn init []
+  (rf/reg-event-fx
+   ::save-auth-token
+   save-auth-token)
+
+  (rf/reg-event-fx
+   ::sign-out
+   sign-out-event))
+
+(defn load-gapi []
```
Comment (kirang89) [inline src/cljs/doit/auth.cljs]: Why is this not called inside `auth/init` ?
